### ISSUE TITLE
Enhancement: [@typescript-eslint/init-declarations] ignore undefined

### ISSUE THREAD

**Issue by rostislav-simonik** [typescript-eslint/typescript-eslint#9950]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/init-declarations/

### Description

We would like to relax the rule to allow declaration without initialization for variables that allow undefined type. cc: @mightyiam 

### Fail

```TypeScript
N/A (we are suggesting relaxation)
```

### Pass

```TypeScript
let foo: string | undefined;
```

### Additional Info

A possible workaround is `let foo: string | undefined = undefined` but the rule `no-undef-init` disallows that.

**Comment by bradzacher** [comment]
The rule is designed to always enforce the style that variables are always initialised and intentionally does not have any exceptions.

I don't think this change aligns with the core rule. It sounds to me like the rule just doesn't align with the code style in your codebase.

---

> but the rule no-undef-init disallows that.

That is just a case of two rules that are incompatible together, right? How can you initialise every variable when a valid value isn't allowed at all? The only way to avoid using `undefined` is either to not use it (and violate `init-declarations`) or use something else (eg `null`).

**Comment by mightyiam** [comment]
Hrm... I changed my mind and I tend to agree. This suggested exception contradicts the intention of the rule.

Thank you.
